bin/cml-tensorboard-dev.js

```diff
@@ -32,51 +33,52 @@ const run = async (opts) => {
   await fs.writeFile(`${path}/uploader-creds.json`, credentials);
 
   // launch  tensorboard on background
-  const tb_path = await exec('which tensorboard');
   const help = await exec('tensorboard dev upload -h');
   const extra_params_found =
     (name || description) && help.indexOf('--description') >= 0;
   const extra_params = extra_params_found
     ? `--name "${name}" --description "${description}"`
     : '';
-  const command = `python -u ${tb_path} dev upload --logdir ${logdir} ${extra_params}`;
 
-  const proc = spawn(command, { detached: true, shell: true });
-
-  proc.stderr.on('data', (data) => {
-    data && console.error(data.toString('utf8'));
+  const command = `tensorboard dev upload --logdir ${logdir} ${extra_params}`;
+  const stdout_path = tempy.file({ extension: 'log' });
+  const stdout_fh = await fs.open(stdout_path, 'a');
+  const stderr_path = tempy.file({ extension: 'log' });
+  const stderr_fh = await fs.open(stderr_path, 'a');
+  const proc = spawn(command, [], {
+    detached: true,
+    shell: true,
+    stdio: ['ignore', stdout_fh, stderr_fh]
   });
 
-  let output = '';
-  proc.stdout.on('data', async (data) => {
-    if (data) {
-      console.error(data.toString('utf8'));
+  proc.unref();
+  proc.on('exit', (code) => {
+    throw new Error(`Tensorboard process exited with code ${code}`);
+  });
 
-      output += data.toString('utf8');
+  // reads stdout every 5 secs to find the tb uri
+  setInterval(async () => {
+    const stdout_data = await fs.readFile(stdout_path, 'utf8');
+    const regex = /(https?:\/\/[^\s]+)/;
+    const matches = stdout_data.match(regex);
 
-      const uri_index = output.indexOf('https://', 0);
-      if (uri_index > -1) {
-        output = output.substring(uri_index, output.length - 1);
+    if (matches.length) {
+      let output = matches[0];
 
-        if (md) output = `[${title || name}](${output})`;
+      if (md) output = `[${title || name}](${output})`;
 
-        if (!file) print(output);
-        else await fs.writeFile(file, output);
+      if (!file) print(output);
+      else await fs.appendFile(file, output);
 
-        process.exit(0);
-      }
+      stdout_fh.close() && stderr_fh.close();
+      process.exit(0);
     }
-  });
-
-  proc.on('exit', (code) => {
-    console.error(output);
-    throw new Error(`Tensorboard process exited with code ${code}`);
-  });
-
-  proc.unref();
+  }, 1 * 5 * 1000);
 
-  setTimeout(() => {
-    // waits 1 min before dies
+  // waits 1 min before dies
+  setTimeout(async () => {
+    stdout_fh.close() && stderr_fh.close();
```

Comment:
Is `&&` required?  What if `stdout_fh.close()` fails? Should we keep `stderr_fh`open?

Reply:
both fh should exist at this point. If not the error should have happened before. If we are not closing it node js will do anyway.

Reply:
It might be a case when the 1st return an error due to system error - hard drive or OS issues.
The condition does not look safe.

Reply:
I'm talking about the case when `stdout_fh.close()` returns an error due to a system issue and we do not close `stderr_fh` that has to be closed.

Reply:
wrapped in try every close